### PR TITLE
Add parallel worlds and counterfactual graphs

### PR DESCRIPTION
**Pull request by cthoyt**

### CODE REVIEW

File: src/y0/graph.py [hunk 1]
```diff
@@ -415,6 +423,29 @@ def is_connected(self) -> bool:
...
+        :returns: A graph that has been intervened on the given interventions
+        """
+        variables = _upgrade_variables(variables)
+        bases = {variable.get_base() for variable in variables}
+        return self.from_edges(
+            nodes=[node.intervene(variables) for node in self.nodes()],
+            directed=[
+                (u.intervene(variables), v.intervene(variables))
+                for u, v in self.directed.edges()
+                # TODO document why not check `u.get_base() not in bases`
```
Comment (cthoyt) [inline src/y0/graph.py:439]: when we intervene, we want to make sure there are no incoming edges into the variables that are getting intervened on

File: src/y0/algorithm/identify/cg.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+
+def _check_world(*, world: World, node: Variable) -> bool:
+    return (node not in world) and (~node not in world)
+
+
+def _stitch_1(
+    graph: NxMixedGraph, worlds: Worlds
+) -> List[Tuple[CounterfactualVariable, CounterfactualVariable]]:
+    # TODO high level documentation
```
Comment (cthoyt) [inline src/y0/algorithm/identify/cg.py:230]: tie together factual world and counterfactual world

**Comment by djinnome** [comment]
All tests pass.  Sufficient coverage for merge to main

**Comment by djinnome** [comment]
Whoops. Didn't auto-squash and merge. Waiting for tests to pass before closing...

**Comment by djinnome** [comment]
Thanks to @njmmerrill we have cleaned up  all flake8 errors and are ready to merge to main. Huzzah!

**Pull request merged**